### PR TITLE
Fix SerialPort memory leak when requests time out and no data is received

### LINKED ISSUE #55146: Possible SerialPort memory leak on Raspbian (all Unix?)
### Description

When using SerialPort.Read(byte[] buffer, int offset, int count) and a TimeoutException occures, the reference for the byte[] buffer used is not released by the underlying SerialStream used by SerialPort.

Sample repo: https://github.com/dseeg/SerialPortMemoryLeak

Application runs though 20000 iterations of trying to read from a SerialPort and timing out each time.

### Configuration

- **.NET Version:** .NET 5.0.301
- **OS:** Raspbian 9.13 stretch
- **Architecture:** ARM

### Regression?

Unknown

#### dotnet-counters
```
Press p to pause, r to resume, q to quit.
    Status: Running

[System.Runtime]
    % Time in GC since last GC (%)                                 0
    Allocation Rate (B / 1 sec)                               33,260
    CPU Usage (%)                                                  1
    Exception Count (Count / 1 sec)                                0
    GC Fragmentation (%)                                           7.642
    GC Heap Size (MB)                                             24
    Gen 0 GC Count (Count / 1 sec)                                 1
    Gen 0 Size (B)                                             9,884
    Gen 1 GC Count (Count / 1 sec)                                 0
    Gen 1 Size (B)                                           485,444
    Gen 2 GC Count (Count / 1 sec)                                 0
    Gen 2 Size (B)                                        25,744,976
    IL Bytes Jitted (B)                                       75,380
    LOH Size (B)                                             131,120
    Monitor Lock Contention Count (Count / 1 sec)                  0
    Number of Active Timers                                        0
    Number of Assemblies Loaded                                   19
    Number of Methods Jitted                                     735
    POH (Pinned Object Heap) Size (B)                         20,016
    ThreadPool Completed Work Item Count (Count / 1 sec)           0
    ThreadPool Queue Length                                        0
    ThreadPool Thread Count                                        0
    Working Set (MB)                                              84
```

#### dumpheap -stat
```
...
6604d32c       11       262276 System.Collections.Concurrent.ConcurrentQueueSegment`1+Slot[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]][]
6a2c3248    20000       560000 System.IO.Ports.SerialStream+SerialStreamIORequest
6601239c    20000       800000 System.Threading.Tasks.Task`1[[System.Int32, System.Private.CoreLib]]
66014c18    20000       880000 System.Threading.Tasks.Task+ContingentProperties
66012008    20002       880088 System.Threading.CancellationTokenSource
011b9e40    31952      2024484      Free
6a2b700c    20025     20732431 System.Byte[]
Total 135477 objects
```

#### gcroot inspection
```
...
53bf10f4 6a2b700c     1036
53bf157c 6a2b700c     1036
53bf1a04 6a2b700c     1036
53bf1e8c 6a2b700c     1036
53bf2314 6a2b700c     1036
53bf27a8 6a2b700c     1036
53bf2c3c 6a2b700c     1036
53bf30c4 6a2b700c     1036
53bf354c 6a2b700c     1036
53bf39f8 6a2b700c     1036
53bf3eb4 6a2b700c     1036
53bf5708 6a2b700c     1036
53bf6a70 6a2b700c     1036
53bfa4a0 6a2b700c     1036
53bfce58 6a2b700c     1036
53c01688 6a2b700c     1036
53c08bcc 6a2b700c     1036

Statistics:
      MT    Count    TotalSize Class Name
6a2b700c    20025     20732431 System.Byte[]
Total 20025 objects

> gcroot 53bfa4a0
Found 0 unique roots (run 'gcroot -all' to see all roots).
> gcroot 53bf30c4
Thread 75d4:
    7E9016F0 6A215A8C SerialPortMemoryLeak.Program.Main(System.String[]) [C:\Users\Dillon\source\repos\SerialPortMemoryLeak\SerialPortMemoryLeak\Program.cs @ 55]
        r11-44: 7e90171c
            ->  62F08D38 System.IO.Ports.SerialPort
            ->  62F09288 System.IO.Ports.SerialStream
            ->  62F09310 System.Collections.Concurrent.ConcurrentQueue`1[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]]
            ->  53766868 System.Collections.Concurrent.ConcurrentQueueSegment`1[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]]
            ->  63F00020 System.Collections.Concurrent.ConcurrentQueueSegment`1+Slot[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]][]
            ->  53BF3508 System.IO.Ports.SerialStream+SerialStreamIORequest
            ->  53BF30C4 System.Byte[]

Found 1 unique roots (run 'gcroot -all' to see all roots).
> gcroot 53bf1e8c
Thread 75d4:
    7E9016F0 6A215A8C SerialPortMemoryLeak.Program.Main(System.String[]) [C:\Users\Dillon\source\repos\SerialPortMemoryLeak\SerialPortMemoryLeak\Program.cs @ 55]
        r11-44: 7e90171c
            ->  62F08D38 System.IO.Ports.SerialPort
            ->  62F09288 System.IO.Ports.SerialStream
            ->  62F09310 System.Collections.Concurrent.ConcurrentQueue`1[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]]
            ->  53766868 System.Collections.Concurrent.ConcurrentQueueSegment`1[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]]
            ->  63F00020 System.Collections.Concurrent.ConcurrentQueueSegment`1+Slot[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]][]
            ->  53BF22D0 System.IO.Ports.SerialStream+SerialStreamIORequest
            ->  53BF1E8C System.Byte[]

Found 1 unique roots (run 'gcroot -all' to see all roots).
```

### Analysis

From the data I collected using dotnet-counters and dotnet-dump, and inspecting the code for [SerialStream.Unix.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs), the byte[] buffer eventually makes its way into a [SerialStreamIORequest and loaded into a queue](https://github.com/dotnet/runtime/blob/607f98c888b31566c773712c3153236c7cec05d2/src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs#L435). SerialStreamIORequests [should be removed from that queue](https://github.com/dotnet/runtime/blob/607f98c888b31566c773712c3153236c7cec05d2/src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs#L789) when they are marked completed. It seems that either this code isn't being reached or IsCompleted is not being set when the cancellation token fires.

### PR DESCRIPTION
**Pull request by bklop**
Fixes #55146

This also replaces the use of ConcurrentQueue with a synchronized Queue to ensure that completed requests (and related data, including any user-supplied buffers) are immediately eligible for GC. This is very important when using SerialPort on resource constrained IoT devices. Since the most common scenario is that of a single client thread calling Read and/or Write and waiting for the result to come back, the coarse locking on the queue should lead to any contention or performance problems in practice.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=68014) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=68014) before we can accept your contribution.<br/><br/>:x: bklop [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=68014)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=68014) it.</sub>

**Comment by kasperk81** [comment]
did you notice improvement in cpu usage too?  (https://github.com/dotnet/runtime/issues/2379)

File: src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs
Comment (krwq) [inline src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs:1086]: honestly this sounds more like a bug with ConcurrentQueue implementation, could you file an issue about this? If we get confirmation this is a bug then we can temporarily use workaround until issue is fixed

Comment (krwq) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs:1086]: also I'm ok with using queue and locks directly in the code because at times it makes sense to put lock around larger piece of code than on each operation individually (i.e. RemoveCompletedTasks could be locked entirely)

Comment (kasperk81) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs:1086]: it was discussed in https://github.com/dotnet/runtime/issues/28953

**Comment by bklop** [comment]
> did you notice improvement in cpu usage too? (#2379)

I don't think this patch will improve CPU usage in any way (even though I haven't tested it).

But I've also had problems with high CPU usage when calling SerialPort.ReadByte() in a loop. Instead I now read directly from the BaseStream and buffer the data read in my own code. This significantly reduced the CPU load.

File: src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs
Comment (krwq) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs:1086]: I'd still consider using locks directly though as I think that can be done much more efficiently than lock per operation

Comment (bklop) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs:1086]: I can have a look at that this week and update the PR :)

**Comment by danmoseley** [comment]
@bklop are you planning more work here or are you waiting on review?

**Review by krwq — APPROVED** [review]

**Comment by danmoseley** [comment]
@krwq can we merge, or do you plan to do manual tests or something?

**Comment by bklop** [comment]
> @bklop are you planning more work here or are you waiting on review?

I was going to make some changes to the queue locking as suggested by @krwq but unfortunately life happened and other things got in the way :(. But I will have some time the coming week so I will try to make these changes ASAP.

File: src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs
Comment (bklop) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs]: Finally updated this PR, now locks are used directly in the code

**Comment by bklop** [comment]
I've now pushed my latest changes but I noticed that System.IO.Ports.Tests.SerialStream tests are not run as part of the CI build. I assume this is because of  `ConditionalFact(nameof(HasOneSerialPort))`?

Unfortunately I don't have a serial port on my development machine so I can't run it locally either. Are these type of tests run somewhere else? Otherwise I can run some (informal) tests with a real serial port on a linux IOT device next week.

**Comment by bklop** [comment]
@danmoseley / @krwq I tested basic reading/writing with a real serial port on a linux IoT device, works fine. Is there anything more I need to do to get this merged? :)

File: src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs
Comment (krwq) [inline src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs:1015]: I'll recommend to just make it instance method

Comment (krwq) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs:1015]: actually nvm, you're using different lock object

**Review by krwq — APPROVED** [review]
LGTM we can merge once CI is green assuming you've manually done some basic testing. Unfortunately I don't have any setup handy to test this

**Comment by krwq** [comment]
note that if we want this change to be included in 7.0 release we have to finish testing by this Friday (Aug 12). Later this needs to go through separate approval process and be double merged in 7.0 branch and main (it's not horrible if we miss 7.0 since this is separate nuget package but will definitely be easier if we do)

**Comment by bklop** [comment]
It'd be great if this was in the 7.0 release (we currently have a workaround where we dispose and recreate the SerialPort instance to avoid memory leaks on a timeout, but we'd like to get rid of this). Can this be merged, or do you need to do more testing yourself?

**Comment by krwq** [comment]
@bklop thanks, try out the nightly package whenever available and validate this fixes your issue

**Pull request merged**
